File: src/utils/pda.rs
use solana_program::pubkey::Pubkey;

pub fn get_wrapper_token_mint(
    token_2022_mint: Pubkey,
    program_id: Pubkey,
) -> (Pubkey, u8, Vec<Vec<u8>>) {
    let (addr, bump) =
        Pubkey::find_program_address(&[b"wrapper", token_2022_mint.as_ref()], &program_id);

    let seeds = vec![
        b"wrapper".to_vec(),
        token_2022_mint.as_ref().to_vec(),
        vec![bump],
    ];

    (addr, bump, seeds)
}

pub fn get_reserve_authority(
    token_2022_mint: Pubkey,
    program_id: Pubkey,
) -> (Pubkey, u8, Vec<Vec<u8>>) {
    let (addr, bump) = Pubkey::find_program_address(
        &[b"reserve_authority", token_2022_mint.as_ref()],
        &program_id,
    );

    let seeds = vec![
        b"reserve_authority".to_vec(),
        token_2022_mint.as_ref().to_vec(),
        vec![bump],
    ];

    (addr, bump, seeds)
}

pub fn get_reserve_authority_token_account(
    token_2022_mint: Pubkey,
    owner: Pubkey,
    program_id: Pubkey,
) -> (Pubkey, u8, Vec<Vec<u8>>) {
    let (addr, bump) = Pubkey::find_program_address(
        &[
            b"reserve_authority_token_account",
            token_2022_mint.as_ref(),
            owner.as_ref(),
        ],
        &program_id,
    );

    let seeds = vec![
        b"reserve_authority_token_account".to_vec(),
        token_2022_mint.as_ref().to_vec(),
        owner.as_ref().to_vec(),
        vec![bump],
    ];

    (addr, bump, seeds)
}

pub fn get_token_mint_authority(
    token_mint: Pubkey,
    program_id: Pubkey,
) -> (Pubkey, u8, Vec<Vec<u8>>) {
    let (addr, bump) =
        Pubkey::find_program_address(&[b"mint_authority", token_mint.as_ref()], &program_id);

    let seeds: Vec<Vec<u8>> = vec![
        b"mint_authority".to_vec(),
        token_mint.as_ref().to_vec(),
        vec![bump],
    ];

    (addr, bump, seeds)
}

pub fn get_token_freeze_authority(
    token_mint: Pubkey,
    program_id: Pubkey,
) -> (Pubkey, u8, Vec<Vec<u8>>) {
    let (addr, bump) =
        Pubkey::find_program_address(&[b"freeze_authority", token_mint.as_ref()], &program_id);

    let seeds: Vec<Vec<u8>> = vec![
        b"freeze_authority".to_vec(),
        token_mint.as_ref().to_vec(),
        vec![bump],
    ];

    (addr, bump, seeds)
}

File: sdk/src/utils/pda.ts
import * as web3 from "@solana/web3.js";
import {
  PROGRAM_ID,
  RESERVE_AUTHORITY_SEED,
  RESERVE_AUTHORITY_TOKEN_ACCOUNT_SEED,
} from "../constants";
import { WRAPPER_TOKEN_MINT_SEED } from "../constants";

export const getWrapperTokenMint = (
  token2022Mint: web3.PublicKey
): web3.PublicKey => {
  const [wrapperTokenMint, _] = web3.PublicKey.findProgramAddressSync(
    [WRAPPER_TOKEN_MINT_SEED, token2022Mint.toBuffer()],
    PROGRAM_ID
  );

  return wrapperTokenMint;
};

export const getReserveAuthority = (
  token2022Mint: web3.PublicKey
): web3.PublicKey => {
  const [reserveAuthority, _] = web3.PublicKey.findProgramAddressSync(
    [RESERVE_AUTHORITY_SEED, token2022Mint.toBuffer()],
    PROGRAM_ID
  );

  return reserveAuthority;
};

export const getReserveAuthorityTokenAccount = (
  token2022Mint: web3.PublicKey
): web3.PublicKey => {
  let reserveAuthority = getReserveAuthority(token2022Mint);

  const [reserveAuthorityTokenAccount, _] =
    web3.PublicKey.findProgramAddressSync(
      [
        RESERVE_AUTHORITY_TOKEN_ACCOUNT_SEED,
        token2022Mint.toBuffer(),
        reserveAuthority.toBuffer(),
      ],
      PROGRAM_ID
    );

  return reserveAuthorityTokenAccount;
};
